File: benchmarks/go/medium.go
package main

import "fmt"

type Point struct {
	X float64
	Y float64
}

func newPoint(x, y float64) Point {
	return Point{X: x, Y: y}
}

func distance(a, b Point) float64 {
	dx := a.X - b.X
	dy := a.Y - b.Y
	// manual sqrt via Newton's method
	val := dx*dx + dy*dy
	if val == 0 {
		return 0
	}
	guess := val / 2
	for i := 0; i < 20; i++ {
		guess = (guess + val/guess) / 2
	}
	return guess
}

func buildSlice(n int) []int {
	result := make([]int, 0, n)
	for i := 0; i < n; i++ {
		result = append(result, i*i)
	}
	return result
}

func filterEven(nums []int) []int {
	result := make([]int, 0)
	for _, v := range nums {
		if v%2 == 0 {
			result = append(result, v)
		}
	}
	return result
}

func sumSlice(nums []int) int {
	total := 0
	for _, v := range nums {
		total += v
	}
	return total
}

func bubbleSort(nums []int) []int {
	sorted := make([]int, len(nums))
	copy(sorted, nums)
	n := len(sorted)
	for i := 0; i < n; i++ {
		for j := 0; j < n-i-1; j++ {
			if sorted[j] > sorted[j+1] {
				sorted[j], sorted[j+1] = sorted[j+1], sorted[j]
			}
		}
	}
	return sorted
}

func classify(n int) string {
	switch {
	case n < 0:
		return "negative"
	case n == 0:
		return "zero"
	case n < 10:
		return "small"
	case n < 100:
		return "medium"
	default:
		return "large"
	}
}

func makeAdder(x int) func(int) int {
	return func(y int) int {
		return x + y
	}
}

func formatPoint(p Point) string {
	return fmt.Sprintf("(%.2f, %.2f)", p.X, p.Y)
}

func main() {
	p1 := newPoint(3.0, 4.0)
	p2 := newPoint(0.0, 0.0)
	fmt.Println(formatPoint(p1))
	fmt.Println(formatPoint(p2))
	fmt.Printf("distance: %.4f\n", distance(p1, p2))

	nums := buildSlice(20)
	evens := filterEven(nums)
	fmt.Println("sum of squares:", sumSlice(nums))
	fmt.Println("sum of even squares:", sumSlice(evens))

	unsorted := []int{64, 34, 25, 12, 22, 11, 90}
	sorted := bubbleSort(unsorted)
	fmt.Println("sorted:", sorted)

	for _, v := range []int{-5, 0, 3, 42, 999} {
		fmt.Printf("%d is %s\n", v, classify(v))
	}

	add5 := makeAdder(5)
	add10 := makeAdder(10)
	fmt.Println("add5(3):", add5(3))
	fmt.Println("add10(7):", add10(7))
}

File: benchmarks/rust/medium.rs
struct Point {
    x: f64,
    y: f64,
}

impl Point {
    fn new(x: f64, y: f64) -> Point {
        Point { x, y }
    }

    fn distance(&self, other: &Point) -> f64 {
        let dx = self.x - other.x;
        let dy = self.y - other.y;
        (dx * dx + dy * dy).sqrt()
    }

    fn format(&self) -> String {
        format!("({:.2}, {:.2})", self.x, self.y)
    }
}

fn build_vec(n: usize) -> Vec<i64> {
    let mut v = Vec::new();
    for i in 0..n {
        v.push(i as i64);
    }
    v
}

fn filter_even(v: &[i64]) -> Vec<i64> {
    let mut result = Vec::new();
    for &x in v {
        if x % 2 == 0 {
            result.push(x);
        }
    }
    result
}

fn sum(v: &[i64]) -> i64 {
    let mut total: i64 = 0;
    for &x in v {
        total += x;
    }
    total
}

fn bubble_sort(v: &mut Vec<i64>) {
    let n = v.len();
    for i in 0..n {
        for j in 0..n - 1 - i {
            if v[j] > v[j + 1] {
                v.swap(j, j + 1);
            }
        }
    }
}

fn describe_number(n: i64) -> &'static str {
    match n % 4 {
        0 => "divisible by 4",
        1 => "remainder 1",
        2 => "remainder 2",
        3 => "remainder 3",
        _ => "unknown",
    }
}

fn make_adder(base: i64) -> Box<dyn Fn(i64) -> i64> {
    Box::new(move |x| base + x)
}

fn format_stats(name: &str, count: usize, total: i64) -> String {
    let avg = if count > 0 {
        total as f64 / count as f64
    } else {
        0.0
    };
    format!("{}: count={}, sum={}, avg={:.2}", name, count, total, avg)
}

fn process_range(start: i64, end: i64) -> Vec<i64> {
    let mut result = Vec::new();
    let mut i = start;
    while i < end {
        if i % 3 != 0 {
            result.push(i * i);
        }
        i += 1;
    }
    result
}

fn main() {
    let p1 = Point::new(0.0, 0.0);
    let p2 = Point::new(3.0, 4.0);
    println!(
        "Distance from {} to {}: {:.2}",
        p1.format(),
        p2.format(),
        p1.distance(&p2)
    );

    let v = build_vec(20);
    let evens = filter_even(&v);
    let total = sum(&evens);
    println!("{}", format_stats("evens", evens.len(), total));

    let mut to_sort = vec![9, 3, 7, 1, 8, 2, 6, 4, 5, 0];
    bubble_sort(&mut to_sort);
    println!("Sorted: {:?}", to_sort);

    for i in 0..8 {
        println!("{}: {}", i, describe_number(i));
    }

    let add5 = make_adder(5);
    let add10 = make_adder(10);
    println!("add5(3) = {}, add10(3) = {}", add5(3), add10(3));

    let squares = process_range(1, 15);
    println!("Squares (non-mult-3): {:?}", squares);
    println!("Sum of squares: {}", sum(&squares));
}
